src/circom/r1cs_reader.rs

```diff
@@ -54,21 +57,39 @@ impl<E: PairingEngine> R1CSFile<E> {
             return Err(Error::new(ErrorKind::InvalidData, "Unsupported version"));
         }
 
-        let _num_sections = reader.read_u32::<LittleEndian>()?;
+        let num_sections = reader.read_u32::<LittleEndian>()?;
 
-        // todo: rewrite this to support different section order and unknown sections
         // todo: handle sec_size correctly
-        let _sec_type = reader.read_u32::<LittleEndian>()?;
-        let sec_size = reader.read_u64::<LittleEndian>()?;
+        // section type -> file offset
+        let mut sec_offsets = HashMap::<u32, u64>::new();
+        let mut sec_sizes = HashMap::<u32, u64>::new();
+
+        // get file offset of each section
+        for _ in 0..num_sections {
```

Comment:
Would be cool if this could be an iterator so that we could then introduce rayon and make the parsing faster

Reply:
Like this? https://github.com/gakonst/ark-circom/pull/13/commits/bc88c726590e250c2e042e9d4b74b77e294e32ec The vec init seems a bit odd to me but maybe this is idiomatic, vs putting `for 0..num_inputs`.

I notice `rayon` is a transitive dependency (Cargo.lock), but that `par_iter` isn't currently used. If we want to introduce this later on, it seems like we'd just have to replace `iter` with `par_iter` https://github.com/rayon-rs/rayon#parallel-iterators-and-more

Reply:
Ah no, I meant you'd do iter().map(...) or iter().fold(...) to create the 2 hash maps instead of using a loop at all!

In any case, bit of a nit, let's revert that change and just do 0..num and we'll make it faster if needed later